[SelectionDAG] Use unaligned store/load to move AVX registers onto stack for `insertelement`

Nirhar opened the pull request:
Prior to this patch, SelectionDAG generated aligned move onto stacks for AVX registers when the function was marked as a no-realign-stack function. This lead to misalignment between the stack and the instruction generated. This patch fixes the issue. There was a similar issue reported for `extractelement` which was fixed in a6614ec5b7c1dbfc4b847884c5de780cf75e8e9c

Nirhar commented:
@RKSimon @danilaml @arsenm for your reviews

Nirhar commented:
We seem to have 2 different functions that essentially do the same job of spilling the vector onto the stack for inserting into it, which are: 
https://github.com/llvm/llvm-project/blob/f81f7a57664bd7dccd5e179ed01155efbc8ff481/llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp#L125
and
https://github.com/llvm/llvm-project/blob/f81f7a57664bd7dccd5e179ed01155efbc8ff481/llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp#L193

As a suggestion for a different PR, I think we can get rid of one of them, or call one through the other.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -363,6 +363,19 @@ SDValue SelectionDAGLegalize::ExpandConstant(ConstantSDNode *CP) {
+  uint64_t ObjectSize = isObjectScalable ? ~UINT64_C(0) : MFI.getObjectSize(FI);
+  return MF.getMachineMemOperand(PtrInfo, MachineMemOperand::MOStore,
+                                 ObjectSize, MFI.getObjectAlign(FI));
+}
```
Move this method up as a NFC pre-commit to reduce the diff

Nirhar replied:
Separated this into a different PR: https://github.com/llvm/llvm-project/pull/82171

RKSimon commented:
ExpandInsertToVectorThroughStack is for INSERT_SUBVECTOR not INSERT_VECTOR_ELT, but I agree we can probably share more of the code between them.

Nirhar commented:
We have about 10 Aarch64 Unit tests failing, all of them due to the following assertion failure:
```
AArch64Tests: /home/nirhar/Documents/OpenSource/llvm/llvm-project/llvm/include/llvm/CodeGen/GlobalISel/LegalizerInfo.h:576: void llvm::LegalizeRuleSet::aliasTo(unsigned int): Assertion `Rules.empty() && "Aliasing will discard rules"' failed.
 #0 0x00007feec86b1188 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/nirhar/Documents/OpenSource/llvm/llvm-project/llvm/lib/Support/Unix/Signals.inc:723:22
 #1 0x00007feec86b15a9 PrintStackTraceSignalHandler(void*) /home/nirhar/Documents/OpenSource/llvm/llvm-project/llvm/lib/Support/Unix/Signals.inc:798:1
 #2 0x00007feec86aeb49 llvm::sys::RunSignalHandlers() /home/nirhar/Documents/OpenSource/llvm/llvm-project/llvm/lib/Support/Signals.cpp:105:20
 #3 0x00007feec86b0a20 SignalHandler(int) /home/nirhar/Documents/OpenSource/llvm/llvm-project/llvm/lib/Support/Unix/Signals.inc:413:1
 #4 0x00007feec7c42520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #5 0x00007feec7c969fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)
 #6 0x00007feec7c42476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #7 0x00007feec7c287f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #8 0x00007feec7c2871b (/lib/x86_64-linux-gnu/libc.so.6+0x2871b)
 #9 0x00007feec7c39e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#10 0x00007feecb9d2d42 llvm::LegalizeRuleSet::aliasTo(unsigned int) /home/nirhar/Documents/OpenSource/llvm/llvm-project/llvm/include/llvm/CodeGen/GlobalISel/LegalizerInfo.h:577:13
#11 0x00007feecb9d16c1 llvm::LegalizerInfo::aliasActionDefinitions(unsigned int, unsigned int) /home/nirhar/Documents/OpenSource/llvm/llvm-project/llvm/lib/CodeGen/GlobalISel/LegalizerInfo.cpp:319:1
#12 0x00007feecb9d15c4 llvm::LegalizerInfo::getActionDefinitionsBuilder(std::initializer_list<unsigned int>) /home/nirhar/Documents/OpenSource/llvm/llvm-project/llvm/lib/CodeGen/GlobalISel/LegalizerInfo.cpp:305:3
#13 0x00007feecc0d6816 llvm::LegalizeRuleSet::legalFor(std::initializer_list<llvm::LLT>) /home/nirhar/Documents/OpenSource/llvm/llvm-project/llvm/include/llvm/CodeGen/GlobalISel/LegalizerInfo.h:600:21
#14 0x00007feecc0d6816 llvm::AArch64LegalizerInfo::AArch64LegalizerInfo(llvm::AArch64Subtarget const&) /home/nirhar/Documents/OpenSource/llvm/llvm-project/llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:126:0
#15 0x00007feecc2b68d6 std::__uniq_ptr_impl<llvm::LegalizerInfo, std::default_delete<llvm::LegalizerInfo>>::reset(llvm::LegalizerInfo*) /usr/include/c++/11/bits/unique_ptr.h:179:16
#16 0x00007feecc2b68d6 std::unique_ptr<llvm::LegalizerInfo, std::default_delete<llvm::LegalizerInfo>>::reset(llvm::LegalizerInfo*) /usr/include/c++/11/bits/unique_ptr.h:456:12
#17 0x00007feecc2b68d6 llvm::AArch64Subtarget::AArch64Subtarget(llvm::Triple const&, llvm::StringRef, llvm::StringRef, llvm::StringRef, llvm::TargetMachine const&, bool, unsigned int, unsigned int, bool, bool, bool) /home/nirhar/Documents/OpenSource/llvm/llvm-project/llvm/lib/Target/AArch64/AArch64Subtarget.cpp:333:0
#18 0x000055f3624367b3 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>::_M_data() const /usr/include/c++/11/bits/basic_string.h:195:28
#19 0x000055f3624367b3 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>::_M_is_local() const /usr/include/c++/11/bits/basic_string.h:230:23
#20 0x000055f3624367b3 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>::_M_dispose() /usr/include/c++/11/bits/basic_string.h:239:18
#21 0x000055f3624367b3 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>::~basic_string() /usr/include/c++/11/bits/basic_string.h:672:19
#22 0x000055f3624367b3 (anonymous namespace)::createInstrInfo(llvm::TargetMachine*) /usr/include/c++/11/bits/basic_string.h:671:7
#23 0x000055f362436edf function<InstSizes_STATEPOINT_Test::TestBody()::<lambda(llvm::AArch64InstrInfo&, llvm::MachineFunction&)> > /usr/include/c++/11/bits/std_function.h:437:19
#24 0x000055f362436edf InstSizes_STATEPOINT_Test::TestBody() /home/nirhar/Documents/OpenSource/llvm/llvm-project/llvm/unittests/Target/AArch64/InstSizes.cpp:155:0
#25 0x00007feecbeb09ca testing::Test::Run() /home/nirhar/Documents/OpenSource/llvm/llvm-project/third-party/unittest/googletest/src/gtest.cc:2687:50
#26 0x00007feecbeb09ca testing::Test::Run() /home/nirhar/Documents/OpenSource/llvm/llvm-project/third-party/unittest/googletest/src/gtest.cc:2677:6
#27 0x00007feecbeb0bdd testing::TestInfo::Run() /home/nirhar/Documents/OpenSource/llvm/llvm-project/third-party/unittest/googletest/src/gtest.cc:2836:14
#28 0x00007feecbeb7172 testing::TestSuite::Run() (.part.0) /home/nirhar/Documents/OpenSource/llvm/llvm-project/third-party/unittest/googletest/src/gtest.cc:3017:35
#29 0x00007feecbeb796f testing::internal::UnitTestImpl::RunAllTests() /home/nirhar/Documents/OpenSource/llvm/llvm-project/third-party/unittest/googletest/src/gtest.cc:5921:41
#30 0x00007feecbea9fe1 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>::_M_data() const /usr/include/c++/11/bits/basic_string.h:195:28
#31 0x00007feecbea9fe1 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>::c_str() const /usr/include/c++/11/bits/basic_string.h:2321:23
#32 0x00007feecbea9fe1 testing::internal::ScopedPrematureExitFile::~ScopedPrematureExitFile() /home/nirhar/Documents/OpenSource/llvm/llvm-project/third-party/unittest/googletest/src/gtest.cc:5055:26
#33 0x00007feecbea9fe1 testing::UnitTest::Run() /home/nirhar/Documents/OpenSource/llvm/llvm-project/third-party/unittest/googletest/src/gtest.cc:5489:1
#34 0x00007feecc4c51ca main /home/nirhar/Documents/OpenSource/llvm/llvm-project/third-party/unittest/UnitTestMain/TestMain.cpp:56:1
#35 0x00007feec7c29d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#36 0x00007feec7c29e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#37 0x000055f36242d7d5 _start (/home/nirhar/Documents/OpenSource/llvm/llvm-project/build/unittests/Target/AArch64/./AArch64Tests+0xa7d5)
```

Is there a way to print the rules, so that I can investigate what those rules contain?

arsenm commented:
> We have about 10 Aarch64 Unit tests failing, all of them due to the following assertion failure:
> 
> ```
> AArch64Tests: /home/nirhar/Documents/OpenSource/llvm/llvm-project/llvm/include/llvm/CodeGen/GlobalISel/LegalizerInfo.h:576: void llvm::LegalizeRuleSet::aliasTo(unsigned int): Assertion `Rules.empty() && "Aliasing will discard rules"' failed.


This will be unrelated to this patch

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -398,22 +398,24 @@ SDValue SelectionDAGLegalize::PerformInsertVectorEltInMemory(SDValue Vec,
+                                    DAG.getDataLayout().getPrefTypeAlign(
+                                        VT.getTypeForEVT(*DAG.getContext())));
+
+  return DAG.getLoad(VT, dl, Ch, StackPtr, MachinePointerInfo(),
```
Can't we represent this as FrameIndex base with unknown offset?

Nirhar replied:
Can you please elaborate on what is it that you want to represent this way? My knowledge on this API is still limited, hence I would not have understood your intent.

arsenm replied:
This is a reload of the entire vector, so there isn't any offset to consider. That is, the original MachinePointerInfo was correct. You should be able to use the MachinePointerInfo::getFixedStack as before

Nirhar replied:
Okay, have updated it, please have a look!

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -398,22 +398,24 @@ SDValue SelectionDAGLegalize::PerformInsertVectorEltInMemory(SDValue Vec,
+
+  Align ElementAlignment = std::min(cast<StoreSDNode>(Ch)->getAlign(),
+                                    DAG.getDataLayout().getPrefTypeAlign(
+                                        VT.getTypeForEVT(*DAG.getContext())));
```
Instead of querying the preferred type alignment, this should use the alignment for the real stack object you have. It just happens that today DAG.CreateStackTemporary uses getPrefTypeAlign, but you should check what the MachineFrameInfo says the alignment is for the frame index

Nirhar replied:
Have a look if my modifications are correct

RKSimon commented:
@Nirhar reverse-ping

Nirhar commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -1489,24 +1491,27 @@ SDValue SelectionDAGLegalize::ExpandInsertToVectorThroughStack(SDValue Op) {
+  Align ElementAlignment =
+      std::min(cast<StoreSDNode>(Ch)->getAlign(),
+               DAG.getDataLayout().getPrefTypeAlign(
+                   Op.getValueType().getTypeForEVT(*DAG.getContext())));
```
Have modified it! Have a look

annamthomas commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -1485,21 +1484,24 @@ SDValue SelectionDAGLegalize::ExpandInsertToVectorThroughStack(SDValue Op) {
+  int FI = cast<FrameIndexSDNode>(StackPtr.getNode())->getIndex();
+  MachinePointerInfo PtrInfo =
+      MachinePointerInfo::getFixedStack(DAG.getMachineFunction(), FI);
   return DAG.getLoad(Op.getValueType(), dl, Ch, StackPtr, PtrInfo);
```
You need to pass in the `ElementAlignment` here so that the load uses the one above (which considers the SP alignment).

Nirhar replied:
Thanks for the catch! Forgot it during the last edit

Nirhar commented:
ping @RKSimon @arsenm

arsenm reviewed (commented):
LGTM but It took me a minute to realize what this was doing. The title is misleading. This is not unconditionally using unaligned store, it is avoiding introducing stack realignments

arsenm commented on llvm/test/CodeGen/X86/unaligned-insert-into-vector-through-stack.ll:
```diff
@@ -0,0 +1,18 @@
+  %a = insertelement <8 x i32> %arg1, i32 42, i32 %n
+  ret <8 x i32> %a
+}
+
```
Can you add a few more test cases? and also the test name should be adjusted. Also have side by side cases with different stack alignment values

Nirhar replied:
Added more test cases. Please check @arsenm !

Nirhar commented:
> LGTM but It took me a minute to realize what this was doing. The title is misleading. This is not unconditionally using unaligned store, it is avoiding introducing stack realignments

@arsenm Yeah, I agree, I'll modify it. Just to be on the same note, the intention of this patch is to use unaligned move instructions for AVX registers when the function has the `no-stack-realign` attribute.

Nirhar commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -1469,12 +1469,11 @@ SDValue SelectionDAGLegalize::ExpandInsertToVectorThroughStack(SDValue Op) {
-  MachinePointerInfo PtrInfo =
-      MachinePointerInfo::getFixedStack(DAG.getMachineFunction(), FI);
+  MachineMemOperand *AlignedMMO = getStackAlignedMMO(
+      StackPtr, DAG.getMachineFunction(), VecVT.isScalableVector());
```
I had assumed that it was `getFixedStack` that did the magic of clamping down instruction alignment, but looks like I was wrong. The clamping is done in `CreateStackTemporary` (in the call to MFI::CreateStackObject) here:
https://github.com/llvm/llvm-project/blob/862d837e483437b33f5588f89e62085de3a806b9/llvm/lib/CodeGen/MachineFrameInfo.cpp#L56

I think I need to move the call to `CreateStackTemporary` inside `getStackAlignedMMO`

Nirhar replied:
Please ignore the previous comment. 

The current version of code generates a MachinePointerInfo and then uses the MachinePointerInfo version of the DAG.getStore() API call, which actually ignores StackAlignment completely. This version of the API call only considers the Alignment of the Object, and it looks like it was meant to only consider that. 

When we take the route through getStackAlignedMMO, we transfer the alignment information from MachineFrameInfo to the MachineMemOperand it returns. Now we can use the DAG.getStore MMO API version to maintain the Alignment information. 

@arsenm Hope that answers your question

arsenm replied:
This is a confusing mess of API. I think using an explicit alignment with any getStore would be easier to understand

Nirhar replied:
@arsenm I've modified the patch to only use explicit alignment along with getStore. I'll folllow this patch up with another patch that removes the getStackAlignedMMO function.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -1485,22 +1484,26 @@ SDValue SelectionDAGLegalize::ExpandInsertToVectorThroughStack(SDValue Op) {
-    Ch = DAG.getStore(
-        Ch, dl, Part, SubStackPtr,
-        MachinePointerInfo::getUnknownStack(DAG.getMachineFunction()));
+    Ch = DAG.getStore(Ch, dl, Part, SubStackPtr, AlignedMMO);
```
This needs to apply the offset to the MMO?

Nirhar replied:
There is no need for it in the current version of the patch

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -1485,22 +1484,26 @@ SDValue SelectionDAGLegalize::ExpandInsertToVectorThroughStack(SDValue Op) {
-        Ch, dl, Part, SubStackPtr,
-        MachinePointerInfo::getUnknownStack(DAG.getMachineFunction()),
-        VecVT.getVectorElementType());
+    Ch = DAG.getTruncStore(Ch, dl, Part, SubStackPtr,
```
Ditto?

Nirhar replied:
Same here

Commit:
Consider If StackRelignable during DAG Legalization of `insertelement`

Prior to this patch, SelectionDAG generated aligned move onto stacks for AVX registers
when the function was marked as a no-realign-stack function. This could lead to misalignment
between the stack and the instruction generated. This patch fixes the issue. There was
a similar issue reported for `extractelement` which was fixed in #a6614ec5b7c1dbfc4b847884c5de780cf75e8e9c

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -1474,7 +1474,10 @@ SDValue SelectionDAGLegalize::ExpandInsertToVectorThroughStack(SDValue Op) {
 
   // First store the whole vector.
-  SDValue Ch = DAG.getStore(DAG.getEntryNode(), dl, Vec, StackPtr, PtrInfo);
+  Align ElementAlignment =
```
This is not the element alignment, this is the alignment of the base vector

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -1496,11 +1500,12 @@ SDValue SelectionDAGLegalize::ExpandInsertToVectorThroughStack(SDValue Op) {
         Ch, dl, Part, SubStackPtr,
         MachinePointerInfo::getUnknownStack(DAG.getMachineFunction()),
-        VecVT.getVectorElementType());
+        VecVT.getVectorElementType(), ElementAlignment);
```
I'm still confused about what is going on with the alignment. Can you assert the resulting StoreSDNode's getAlign is the element alignment (the actual element alignment, not the misnamed ElementAlignment here)?

Nirhar replied:
I have added the assert when PartVT is not a vector. If I have to add the same assert when PartVT is a vector, should I compare the alignment of the StoreSDNode with the alignment of the PartVT or with the alignment of an element of the vector?
Also, I hope I am calculating the element alignment correctly. Please let me know if I am not doing so.

Nirhar replied:
Done

RKSimon commented on llvm/test/CodeGen/X86/insert-into-vector-through-stack-no-stack-realign.ll:
```diff
@@ -0,0 +1,122 @@
+  ret <8 x i64> %a
+}
+
+attributes #0 = { "no-realign-stack" "target-cpu"="haswell" }
```
replace target-cpu with -mattr=+avx in the RUN line

RKSimon replied:
also please add "nounwind" to attributes list

Nirhar replied:
Done

RKSimon reviewed (approved):
LGTM
